egs/wsj/s5/utils/prepare_extended_lang.sh

```diff
@@ -0,0 +1,182 @@
+  
+  # Filter lexical entries which are already in the source lexicon
+  awk '{ gsub(/\t/, " "); print }' $extra_lexicon | sort -u | \
+    awk 'NR==FNR{a[$0]=1;next} {if (!($0 in a)) print $0 }' $tmpdir/lexicon.txt - \
+    > $extdict/lexicon_extra.txt || exit 1;
+  
+  echo "$(basename $0): Creating $extdict/lexiconp.txt from $srcdict/lexiconp.txt and $extdict/lexicon_extra.txt"
+  perl -ape 's/(\S+\s+)(.+)/${1}1\t$2/;' < $extdict/lexicon_extra.txt | \
+    cat $srcdict/lexiconp.txt - | sort -u > $extdict/lexiconp.txt || exit 1;
+  
+  perl -ape 's/(\S+\s+)\S+\s+(.+)/$1$2/;' <$extdict/lexiconp.txt  >$extdict/lexicon.txt || exit 1;
+  
+  # Create lexicon_silprobs.txt
+  silprob=false
+  [ -f $srcdict/lexiconp_silprob.txt ] && silprob=true
+  if "$silprob"; then
+    echo "$(basename $0): Creating $extdict/lexiconp_silprob.txt from $srcdict/lexicon_silprob.txt"
+    # Here we assume no acoustic evidence for the extra word-pron pairs.
+    # So we assign silprob1 = overall_silprob, silprob2 = silprob3 = 1.00
+    overall_silprob=`awk '{if ($1=="overall") print $2}' $srcdict/silprob.txt`
+    awk -v overall=$overall_silprob '{
+      printf("%s %d %.1f %.2f %.2f",$1, 1, overall, 1.00, 1.00); 
+      for(n=2;n<=NF;n++) printf " "$n; printf("\n");
+      }' $extdict/lexicon_extra.txt | cat $srcdict/lexiconp_silprob.txt - | \
+      sort -k1,1 -k2g,2 -k6 \
+      > $extdict/lexiconp_silprob.txt || exit 1;
+  fi
+  
+  if ! utils/validate_dict_dir.pl $extdict >&/dev/null; then
+    utils/validate_dict_dir.pl $extdict  # show the output.
+    echo "$(basename $0): Validation failed on the extended dict"
+    exit 1;
+  fi
+fi
+
+if [ $stage -le 1 ]; then
+  echo "$(basename $0): Preparing the extended lang dir."
+  [ -d $extlang ] && rm -r $extlang 2>/dev/null
+  utils/prepare_lang.sh $prep_lang_opts $extdict \
+    $oov_word $tmpdir $extlang || exit 1;
+fi
+
+if [ $stage -le 2 ]; then
+  [ -z "$lm" ] && echo "$(basename $0): Not extending the language model since it's not provided." && exit 0;
+  awk '{print $1}' $extlang/words.txt > $tmpdir/voc || exit 1;
+  
+  echo "$(basename $0): Inserting extra unigrams into '$lm'."
+  gunzip -c $lm | \
+    awk -v s=$extdict/lexicon_extra.txt -v oov=$oov_word -v boost=$oov_unigram_prob -v prob=$oov_prob \
+    'BEGIN{ while((getline < s) > 0) { dict[$1] = 1; a += 1; } unigram = 1; } 
+    { if ($1 == "ngram") { split($2, y, "="); if (y[1] == 1) $2 = y[1]"="y[2]+a; } print $0;
+      if (unigram == 1 && $2 == oov) {
+        prob = log(oov_unigram_prob); 
+        for (x in dict) { $1 = prob; $2 = x; print $0; }
+      } 
+      if ($1 == "\\2-grams:") unigram = 0;
+    }' | change-lm-vocab -vocab $tmpdir/voc -lm - -write-lm - $srilm_opts | gzip -c > $tmpdir/lm_tmp.gz
+
+  if $to_carpa; then
+    echo "$(basename $0): Convert the LM to FST."
+    unk=`cat $extlang/oov.int`
+    bos=`grep -w "<s>" $extlang/words.txt | awk '{print $2}'`
```

Comment:
Better to use something like `awk '{if($1 == "<s>") print $2;}'`, in case there are words like `<s>foo`.